<a href="https://colab.research.google.com/github/tbnorth/LOads_colab/blob/master/LOads.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Setup

```python
# Install the PyDrive wrapper & import libraries.
# This only needs to be done once per notebook.
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client.
# This only needs to be done once per notebook.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
```

In [1]:
import os
from pathlib import Path
from google.colab import drive

drive.mount("/content/gdrive")
ROOT = Path("/content/gdrive/My Drive/Colab Notebooks")
BASE = ROOT.joinpath('LOads')
BASE.mkdir(exist_ok=True)
os.chdir(BASE)


Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [2]:
!apt-get install libspatialindex-dev python3-rtree
!pip install geopandas
!ldconfig
from rtree import index
from rtree.index import Rtree
import numpy as np
import pandas as pd
import geopandas
from collections import namedtuple
from dateutil.parser import parse
import json

Reading package lists... Done
Building dependency tree       
Reading state information... Done
libspatialindex-dev is already the newest version (1.8.5-5).
python3-rtree is already the newest version (0.8.3+ds-1).
The following package was automatically installed and is no longer required:
  libnvidia-common-410
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.


In [0]:
def merge(left, right, on, cols):
    for col in cols:
        if col in left.columns:
            left = left.drop(columns=col)
    right = right.reset_index()
    return left.merge(right[on+cols], on=on, how='left')

In [0]:
# uploaded get_wqdata.py to Colab Notebooks folder, now add to import path
import sys
if BASE not in sys.path:
    sys.path.append(str(BASE))
from get_wqdata import GetWQData, WQ_API_URL

wq = GetWQData()

# Explore dataset

Want to work out which CharacteristicName values to request.  The WQ portal web interface says they come [from here](http://iaspub.epa.gov/sor_internet/registry/substreg/home/overview/home.do) but that's hard to search, so request all obs. in our bounding box and search that list instead.

In [5]:
query = dict(
    _desc="All sample types",
    bBox="-80.0402,43.0646,-75.9697,44.355",
    siteType=["Aggregate surface-water-use", "Stream"],
    sampleMedia=["Water", "water"],
    # siteid=["USGS-04231600"],
    # startDateLo="01-01-2018",
    # startDateHi="02-01-2018",
)
print("Reading data")
all_data = wq.get_data(query)  # gets path to .csv file
d = pd.read_csv(all_data)  # , nrows=10)
print("Data read")
print(d.columns)
po4 = set(d["CharacteristicName"])
keep = set()
for term in 'po4', 'phosphor', 'srp', 'phosphate', 'flow', 'guage', 'cfs':
    for cname in po4:
        if term in cname.lower():
            keep.add(cname)
print(sorted(keep))

Reading data


/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (9,10,11,13,19,20,22,24,25,33,35,39,41,42,44,55,56,57,61) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Data read
Index(['OrganizationIdentifier', 'OrganizationFormalName',
       'ActivityIdentifier', 'ActivityTypeCode', 'ActivityMediaName',
       'ActivityMediaSubdivisionName', 'ActivityStartDate',
       'ActivityStartTime/Time', 'ActivityStartTime/TimeZoneCode',
       'ActivityEndDate', 'ActivityEndTime/Time',
       'ActivityEndTime/TimeZoneCode',
       'ActivityDepthHeightMeasure/MeasureValue',
       'ActivityDepthHeightMeasure/MeasureUnitCode',
       'ActivityDepthAltitudeReferencePointText',
       'ActivityTopDepthHeightMeasure/MeasureValue',
       'ActivityTopDepthHeightMeasure/MeasureUnitCode',
       'ActivityBottomDepthHeightMeasure/MeasureValue',
       'ActivityBottomDepthHeightMeasure/MeasureUnitCode', 'ProjectIdentifier',
       'ActivityConductingOrganizationText', 'MonitoringLocationIdentifier',
       'ActivityCommentText', 'SampleAquifer', 'HydrologicCondition',
       'HydrologicEvent', 'SampleCollectionMethod/MethodIdentifier',
       'SampleCollectionMethod/

Now make a sensible list from the values we found.

In [0]:
use_cache = {
    'characteric_names': [
        # 'O-Ethyl O-methyl S-propyl phosphorothioate',
        # 'O-Ethyl S-methyl S-propyl phosphorodithioate',
        # 'O-Ethyl S-propyl phosphorothioate',
        # 'Organic phosphorus',
        'Orthophosphate',
        'Orthophosphate as P',
        'Phosphate-phosphorus',
        'Phosphate-phosphorus as P',
        'Phosphorus',
        'Stream flow, instantaneous',
        'Stream flow, mean. daily',
        # 'Tributyl phosphate',
        # 'Triphenyl phosphate',
        # 'Tris(1,3-dichloro-2-propyl)phosphate',
        # 'Tris(2-butoxyethyl) phosphate',
        # 'Tris(2-chloroethyl) phosphate',
    ]
}

names = use_cache.get('characteric_names')

Now grab all the data we're interested in

In [7]:
query = dict(
    _desc="Target sample types",
    bBox="-80.0402,43.0646,-75.9697,44.355",
    siteType=["Aggregate surface-water-use", "Stream"],
    sampleMedia=["Water", "water"],
    characteristicName=names,
)
print("Reading data")
# first get the Site info. flavored response
data = wq.get_data(query, type_='site')  # gets path to .csv file
site = pd.read_csv(data)
# print('\n'.join(sorted(site.columns)))
# then get the Result info.
data = wq.get_data(query, type_='result')
d = pd.read_csv(data, low_memory=False)
d.columns = [i.replace('/', '_') for i in d.columns]
# get rid of "mg/l<space><space><space>" units
for col in ['ResultMeasure_MeasureUnitCode']:
    d[col] = [str(i).strip() for i in d[col]]
had = len(d)
d = d.loc[d['ResultMeasure_MeasureUnitCode'] != 'nan', :]
print("Lost %d for missing units" % (had - len(d)))


Reading data
Lost 2251 for missing units


In [8]:
# find common columns
common = set(site.columns).intersection(set(d.columns))
print('\n'.join(sorted(common)))
# use this common column
common = 'MonitoringLocationIdentifier'

MonitoringLocationIdentifier
OrganizationFormalName
OrganizationIdentifier
ProviderName


# Merge sites

In [9]:
# create a list of unique sites
locs = site.loc[:, [common, 'LatitudeMeasure', 'LongitudeMeasure']]
locs.drop_duplicates(inplace=True)
length = len(locs)
length

230

In [0]:
# copy lat / lon to main data
old_len = len(d)
locs.reset_index(inplace=True, drop=True)
d = d.merge(locs.loc[:, [common, 'LatitudeMeasure', 'LongitudeMeasure']], on=[common])
assert len(d) == old_len
assert sum(d.LatitudeMeasure == 0) == 0
assert sum(np.isnan(d.LatitudeMeasure)) == 0

Data has different `MonitoringLocationIdentifier`s for the same coords.

In [11]:
# generate a location based site field
lonlat = d[['LongitudeMeasure', 'LatitudeMeasure']].drop_duplicates()
dd = geopandas.GeoDataFrame(
    lonlat,
    geometry=geopandas.points_from_xy(
        lonlat.LongitudeMeasure, lonlat.LatitudeMeasure
    ),
    crs = {'init' :'epsg:4326'},
)
dd['site'] = dd.LatitudeMeasure.astype(str)+dd.LongitudeMeasure.astype(str)
dd.drop(columns=['LongitudeMeasure', 'LatitudeMeasure'], inplace=True)
# 50 m buffer merges sites within 100 m of each other
dd.geometry = dd.geometry.to_crs(5071).buffer(50)
intersect = geopandas.overlay(dd, dd, how='identity')
lint = len(intersect)
intersect = intersect[intersect.site_2.astype(str) != 'nan']
print("Dropped %d NaN intersections" % (lint-len(intersect)))

# make a dict of site: (set of intersecting sites)
sets = {}
for inter in intersect.itertuples():
    site_1 = str(inter.site_1)
    site_2 = str(inter.site_2)
    sets.setdefault(site_1, set()).add(site_2)
    sets.setdefault(site_2, set()).add(site_1)
# pick the lowest (alphabetically) site as the name for intersecting sites
sites = {min(i):i for i in sets.values()}
# sites = {(k+"_X" if len(v) > 1 else k):v for k, v in sites.items()}
[{k:v} for k,v in sites.items() if len(v) > 1]
len(lonlat), len(sites)
site = {}
xtra = {}
for k, v in sites.items():
    for s in v:
        site[s] = k
        xtra[s] = len(v) > 1

lu = d.loc[:, ['LongitudeMeasure', 'LatitudeMeasure', common]].drop_duplicates(
    subset=['LongitudeMeasure', 'LatitudeMeasure'])
#X lu = merge(lu, d, ['LongitudeMeasure', 'LatitudeMeasure'], [common])
lu['sitehash'] = lu.LatitudeMeasure.astype(str)+lu.LongitudeMeasure.astype(str)
sitename = {k:v for k,v in zip(lu.sitehash, lu[common])}
lu['site'] = [sitename[site[i]]+('_X' if xtra[i] else '') for i in lu.sitehash]
lu.drop_duplicates(inplace=True)

d = merge(d, lu, ['LongitudeMeasure', 'LatitudeMeasure'], ['site'])
assert len(d) == old_len, (old_len, len(d))   

Dropped 4 NaN intersections


Now copy site info. into locs table

In [0]:
nlocs = merge(locs, d, [common, 'LatitudeMeasure', 'LongitudeMeasure'], ['site'])
locs_len = len(nlocs)
nlocs.drop_duplicates(inplace=True, subset=['site', 'MonitoringLocationIdentifier'])
assert len(nlocs) == length, (length, len(nlocs))
locs = nlocs
locs.to_csv("locs.csv", index=False)

In [0]:
for na in '', 'nan', 'NaN':
    assert sum(locs['site'] == na) == 0, na
    assert sum(d['site'] == na) == 0, na
    assert sum(locs['site'] == na+na) == 0, na
    assert sum(d['site'] == na+na) == 0, na

In [14]:
print("%d locations" % len(locs))
distinct = locs.set_index(["LatitudeMeasure", "LongitudeMeasure"])
print("%d distinct" % len(distinct.index.unique()))

230 locations
192 distinct


# Data to sampling events

In [15]:
# tabulate units used for results
d['ResultSampleFractionText'] = [
    i if i != 'nan' else 'NA'
    for i in d['ResultSampleFractionText'].astype(str)
]

res_types = (
    d.groupby(
        [
            'CharacteristicName',
            'ResultMeasure_MeasureUnitCode',
            'ResultSampleFractionText',
        ]
    )
    .count()
    .iloc[:, 0]
)
print(res_types)

CharacteristicName          ResultMeasure_MeasureUnitCode  ResultSampleFractionText
Orthophosphate              mg/l                           Dissolved                     181
                                                           NA                             36
                                                           Total                         119
                            mg/l as P                      Dissolved                   18392
                                                           Total                         380
                            mg/l asPO4                     Dissolved                   18393
Orthophosphate as P         mg/l                           Dissolved                     122
                                                           Total                           1
Phosphate-phosphorus        mg/l                           Total                         417
Phosphate-phosphorus as P   mg/l                           NA                  

In [0]:
duration = wq.db.get('duration', [])  # cache expensive calc. in wq.db
if not duration:
    for row in d.itertuples():
        try:
            t0 = parse(
                "%s %s"
                % (row.ActivityStartDate, row.ActivityStartTime_Time)
            )
            t1 = parse(
                "%s %s" % (row.ActivityEndDate, row.ActivityEndTime_Time)
            )
            duration.append((t1 - t0).seconds)
        except ValueError:
            duration.append(0)
    wq.db['duration'] = duration
d['duration'] = duration

Create an ID field for each sampling event

In [0]:
d['sampling'] = (
    d['ActivityStartTime_Time'].astype(str)
    + d['ActivityEndTime_Time'].astype(str)
    + ' '
    + d['ActivityStartDate'].astype(str)
    + ' '
    + d['ActivityEndDate'].astype(str)
    + ' '
    + d['LatitudeMeasure'].astype(str)
    + ' '
    + d['LongitudeMeasure'].astype(str)
)

## Save output data

In [0]:
d.to_csv('data.csv')

Now we want to match P conc. data and flow data for each sampling event

(start to refer to `d` as `data` from here on)

In [19]:
data = pd.read_csv("data.csv", low_memory=False)
pdata = data.loc[:, ['sampling', 'ActivityStartDate', 'ActivityEndDate', 'site']]
for na in '', 'nan', 'NaN':
    assert sum(locs['site'] == na) == 0, na
    assert sum(d['site'] == na) == 0, na
    assert sum(pdata['site'] == na) == 0, na
    assert sum(locs['site'] == na+na) == 0, na
    assert sum(d['site'] == na+na) == 0, na
    assert sum(pdata['site'] == na+na) == 0, na
pdata.drop_duplicates(subset=['sampling'], inplace=True)
pdata_len = len(pdata)
print("%d into %d" % (len(data), pdata_len))
cname = 'CharacteristicName'
cunit = 'ResultMeasure_MeasureUnitCode'
vname = 'ResultMeasureValue'
TP = namedtuple("ToProcess", "to cname unit trans")
po4_to_p = (30.97 + 4 * 16) / 30.97
to_process = [
    TP('flow', 'Stream flow, instantaneous', 'ft3/s', None),
    TP(
        'flow',
        'Stream flow, instantaneous',
        'm3/sec',
        lambda x: x * 35.3147,
    ),
    TP('flow', 'Stream flow, mean. daily', 'ft3/s', None),
    TP('conc', 'Phosphorus', 'mg/l as P', None),
    TP('conc', 'Phosphorus', 'mg/l', None),
    TP('conc', 'Phosphorus', 'ug/l', lambda x: 1000 * x),
    TP('conc', 'Phosphorus', 'mg/l PO4', lambda x: x / po4_to_p),
    TP('conc', 'Phosphate-phosphorus as P', 'mg/l', None),
    TP('conc', 'Phosphate-phosphorus', 'mg/l', None),
    TP('conc', 'Orthophosphate as P', 'mg/l', None),
    TP('conc', 'Orthophosphate', 'mg/l as P', None),
    TP('conc', 'Orthophosphate', 'mg/l asPO4', lambda x: x / po4_to_p),
    TP('conc', 'Orthophosphate', 'mg/l', lambda x: x / po4_to_p),
    # exclude as it's in bed sediment
    # TP('conc', 'Phosphorus', 'mg/kg as P', None),
]

82386 into 22063


We're merging 82386 observations into 21877 flow + conc. records.  In a perfect world we'd have one flow and one conc. record for each sampling, but we have both redundant and mismatched (flow without conc. and visa versa) records, roughly 4:1 instead of 2:1.

`to_process` is an **ordered** list of items where each item reocords the field we're trying to fill (`to`), the `CharacteristicName` (`cname`) of our prefered source field, our prefered `unit`, and any conversion (`trans`) necessary to use this source field / unit combination.

Now we're going to fill the `to` fields (`flow`, `conc`) with the values selected using the list items, filling as many records with the first set of selected data, and only filling missing records with the next (less desireable) list item.

In [20]:
selected = 0
data.set_index('sampling', drop=False, inplace=True)
pdata.set_index('sampling', drop=False, inplace=True)
for tp_i, tp in enumerate(to_process):
    select = np.logical_and(
        data[cname] == tp.cname, data[cunit] == tp.unit
    )
    selected += sum(select)
    print(
        "%d Selected %d %s %s (%d)"
        % (
            tp_i, sum(select),
            tp.cname,
            tp.unit,
            len(data['sampling'][select].unique()),
        )
    )
    mean = data.loc[select, :]
    mean = mean.groupby(level=0).mean()
    pdata = pdata.join(mean.loc[:, vname])
    assert len(pdata) == pdata_len, (pdata_len, len(pdata))
    if tp.trans is not None:
        pdata[vname] = pdata[vname].map(tp.trans)

    def missing(x, f):
        if f not in x.columns:
            return np.ones(len(x)).astype(bool)
        x = x[f].astype(str)
        ans = np.logical_or(np.equal(x, 'None'), np.equal(x, ''))
        ans = np.logical_or(ans, np.equal(x, 'nan'))
        ans = np.logical_or(ans, np.equal(x, 'NaN'))
        return ans

    missing0 = missing(pdata, tp.to)
    if tp.to not in pdata.columns:
        pdata.rename(columns={vname: tp.to}, inplace=True)
    else:
        # print(len(missing), sum(missing), missing.shape)
        #X pdata[tp.to][missing0] = pdata[vname][missing0]
        pdata.loc[missing0, tp.to] = pdata[vname][missing0]
        pdata.drop(columns=vname, inplace=True)
    missing1 = missing(pdata, tp.to)
    print(
        "%d, Needed %d, used %d, still need %d"
        % (tp_i, sum(missing0), sum(missing0) - sum(missing1), sum(missing1))
    )
    assert len(pdata) == pdata_len, (pdata_len, len(pdata))
print("Selected %s total" % selected)
pdata = pdata[~np.isnan(pdata['conc'])]
pdata = pdata[~np.isnan(pdata['flow'])]
print("Lost %d to blank data" % (pdata_len-len(pdata)))
pdata.to_csv("pdata.csv")

0 Selected 3244 Stream flow, instantaneous ft3/s (3237)
0, Needed 22063, used 3237, still need 18826
1 Selected 3244 Stream flow, instantaneous m3/sec (3237)
1, Needed 18826, used 0, still need 18826
2 Selected 15321 Stream flow, mean. daily ft3/s (15283)
2, Needed 18826, used 15184, still need 3642
3 Selected 21072 Phosphorus mg/l as P (19839)
3, Needed 22063, used 19839, still need 2224
4 Selected 359 Phosphorus mg/l (307)
4, Needed 2224, used 303, still need 1921
5 Selected 3 Phosphorus ug/l (3)
5, Needed 1921, used 3, still need 1918
6 Selected 736 Phosphorus mg/l PO4 (735)
6, Needed 1918, used 128, still need 1790
7 Selected 358 Phosphate-phosphorus as P mg/l (358)
7, Needed 1790, used 358, still need 1432
8 Selected 417 Phosphate-phosphorus mg/l (417)
8, Needed 1432, used 5, still need 1427
9 Selected 123 Orthophosphate as P mg/l (123)
9, Needed 1427, used 0, still need 1427
10 Selected 18772 Orthophosphate mg/l as P (18345)
10, Needed 1427, used 351, still need 1076
11 Selected 

Of 22,000 sampling events sampling either flow or P, we can't find flow *and* P for 4700.  Most (~3600) of the missing data is flow, so we *could* fill in with flow data, not doing that currently.

In [0]:
# update locs with the number of obs. for each site
count = pdata.groupby(by=['site']).count()
count.rename(columns={count.columns[0]:'count'}, inplace=True)
locs = merge(locs, count, ['site'], ['count'])

# as above for min/max date
count = pdata.groupby(by=['site']).min()
locs = merge(locs, count, ['site'], ['ActivityStartDate'])
locs.rename(columns={'ActivityStartDate': 'StartDate'}, inplace=True)
count = pdata.groupby(by=['site']).max()
# yes, use ActivityStartDate both times
locs = merge(locs, count, ['site'], ['ActivityStartDate'])
locs.rename(columns={'ActivityStartDate': 'EndDate'}, inplace=True)

locs.to_csv("locs.csv")

# Link to flow data

Now we need flow data for places without conc. data.  Doesn't seem to be in the WQ portal, not strictly speaking a WQ parameter (well, it's Water Quantity, not Water Quality).

Ran //waterservices.usgs.gov/nwis/site/?format=rdb&bBox=-80.040200,43.064600,-75.969700,44.355000&seriesCatalogOutput=true&siteType=ST&siteStatus=all and created `waterservices.usgs.gov.txt`


In [22]:
floc = pd.read_csv(
    "waterservices.usgs.gov.txt", sep=r'\t', comment='#', engine='python', 
    skiprows=[43],  # the FORTRANesque column widths
    dtype={'site_no': str}
)
print("Total", len(floc))
floc.drop_duplicates(subset=['dec_lat_va', 'dec_long_va', 'agency_cd', 'site_no'], inplace=True)
print("Unique", len(floc))
floc = floc[~np.isnan(floc.dec_lat_va)]
floc = floc[~np.isnan(floc.dec_long_va)]
print("Non-null", len(floc))
floc.to_csv("locs2.csv")

Total 28901
Unique 322
Non-null 321


The P sampling `locs` have their original lat/lon values, no necessarily consistent within a site.  That's ok, use these for intersect as it will give all possible answers for each site, we can resolve mulitples later.

In [23]:
import folium
mdnmap = folium.Map(location=[43.234853, -77.533949], zoom_start=19)

for row in locs.itertuples():
    folium.Marker([row.LatitudeMeasure, row.LongitudeMeasure], popup=row.site).add_to(mdnmap)
mdnmap

# ---
locblobs = geopandas.GeoDataFrame(
    locs,
    geometry=geopandas.points_from_xy(
        locs.LongitudeMeasure, locs.LatitudeMeasure
    ),
    crs = {'init' :'epsg:4326'},
)
locblobs.geometry = locblobs.geometry.to_crs(5071).buffer(50)

# don't do this because we might want to use an up/down stream guage for
# an unguaged location with lots of P obs.
# although this doesn't entirely make sense as we're filtering on the presence
# of at least 1 WQ sample of some kind...
#
# print("From", len(locblobs), "locations")
# locblobs = locblobs[~np.isnan(locblobs['count'])]
# print("Go to", len(locblobs), "with P obs.")

flocblobs = geopandas.GeoDataFrame(
    floc,
    geometry=geopandas.points_from_xy(
        floc.dec_long_va, floc.dec_lat_va
    ),
    crs = {'init' :'epsg:4326'},
)
flocblobs.geometry = flocblobs.geometry.to_crs(5071).buffer(50)
# flocblobs['site_id'] = flocblobs.agency_cd + '-' + flocblobs.site_no
# because geopandas.overlay seems to change site_no to float

intersect = geopandas.overlay(locblobs, flocblobs, how='identity')

floc.site_no.dtype
flocblobs.site_no.dtype
intersect.site_no.dtype
# intersect.site_id.dtype


lint = len(intersect)
intersect = intersect[intersect.site_no.astype(str) != 'nan']
print("Dropped %d NaN intersections" % (lint-len(intersect)))

# make a dict of site: (set of intersecting flow data)
sets = {}
for inter in intersect.itertuples():
    site_1 = str(inter.site)
    site_2 = "%s-%s" % (inter.agency_cd, inter.site_no)
    sets.setdefault(site_1, set()).add(site_2)

[{k:v} for k,v in sets.items() if len(v) > 1]
toget = set()
for flows in sets.values():
    toget.update(flows)

tsv_tmpl = {
    "site_no": 431510077363501,
    "agency_cd": "USGS",
    "inventory_output": "0",
    "rdb_inventory_output": "file",
    "begin_date": "2000-01-01",
    "end_date": "2019-07-11",
    "TZoutput": "0",
    "pm_cd_compare": "Greater than",
    "radio_parm_cds": "all_parm_cds",
    "qw_attributes": "0",
    "format": "rdb",
    "qw_sample_wide": "wide",
    "rdb_qw_attributes": "0",
    "date_format": "YYYY-MM-DD",
    "rdb_compression": "gz",
    "submitted_form": "brief_list",
}
    
    
flows = {}    
for flow in toget:
    flow = flow.replace("USGS-", "")
    # if flow != "04249000":
    #     continue
    query = dict(
        format="json",
        indent="off",
        sites=flow,
        startDT="2000-01-01",
        endDT="2019-07-10",
        parameterCd="00060",  # ft3/s
    )
    jsonfile = wq.get_data(query, 'flow')
    print(WQ_API_URL['flow']+'?'+'&'.join("%s=%s" % (k, v) for k, v in query.items() if not k.startswith('_')))
    data = json.load(open(jsonfile))
    data = data['value']['timeSeries']
    if data:
        data = data[0]['values']
        assert len(data) == 1, len(data)
        data = data[0]['value']
        print(flow, len(data))
        flows[flow] = [(i['dateTime'], float(i['value'])) for i in data]
    else:
        # fall back to TSV data
        query = tsv_tmpl.copy()
        query['site_no'] = str(flow)
        data = pd.read_csv(wq.get_data(query, 'flowTSV'))        

# flows['431510077363501']

Dropped 161 NaN intersections
https://waterservices.usgs.gov/nwis/dv/?format=json&indent=off&sites=04218054&startDT=2000-01-01&endDT=2019-07-10&parameterCd=00060
Query 34c2f20583fdda48e33a1ba473cc05730bcea25dab5c77408e24ceed3f4cce1c
use_get True
{
  "TZoutput": "0",
  "_api": "https://nwis.waterdata.usgs.gov/usa/nwis/qwdata/",
  "agency_cd": "USGS",
  "begin_date": "2000-01-01",
  "date_format": "YYYY-MM-DD",
  "end_date": "2019-07-11",
  "format": "rdb",
  "inventory_output": "0",
  "pm_cd_compare": "Greater than",
  "qw_attributes": "0",
  "qw_sample_wide": "wide",
  "radio_parm_cds": "all_parm_cds",
  "rdb_compression": "gz",
  "rdb_inventory_output": "file",
  "rdb_qw_attributes": "0",
  "site_no": "04218054",
  "submitted_form": "brief_list"
}
Reading data .


OSError: ignored

In [30]:

locs['sFlow'] = None
locs['eFlow'] = None
locs['cFlow'] = None
for loc in locs.itertuples():
    if not loc.count:
        continue
    if loc.site not in sets:
        print(loc.site, "no flow data source found")
        continue
    flow = loc.site.replace("USGS-", "")
    if flow not in flows:
        print(loc.site, "no flow data found")
        continue
    print("Populating", loc.site)
    for src in sets[loc.site]:
        if loc.sFlow:
            print("Overwriting data for", loc.site)
        flow = src.replace("USGS-", "")
        locs.loc[loc.Index, 'sFlow'] = min(i[0] for i in flows[flow])
        locs.loc[loc.Index, 'eFlow'] = max(i[0] for i in flows[flow])
        locs.loc[loc.Index, 'cFlow'] = len(flows[flow])
        # print(len(flows[flow]), 'for', locs.site)
locs.to_csv("locs.csv")
sets['USGS-431510077363501']
# flows['431510077363501']

USGS-04217500 no flow data found
USGS-04218000_X no flow data found
USGS-04218030 no flow data found
USGS-04218035 no flow data found
USGS-04218054 no flow data found
USGS-04218592 no flow data found
USGS-04218610 no flow data found
USGS-04218700 no flow data found
USGS-04218740 no flow data found
USGS-04219355 no flow data found
USGS-04219501 no flow data found
USGS-04219640 no flow data found
USGS-04219650 no flow data found
USGS-04219720 no flow data found
USGS-04219726 no flow data found
USGS-04219735 no flow data found
USGS-04219765 no flow data found
Populating USGS-04219768
USGS-04219775 no flow data found
USGS-04219915 no flow data found
USGS-04219979 no flow data found
USGS-04219997 no flow data found
USGS-0421999720 no flow data found
USGS-0421999750 no flow data found
USGS-04219998 no flow data found
USGS-04219999 no flow data found
USGS-04220045_X no flow data found
Populating USGS-0422016550
Populating USGS-04220223
USGS-04220230 no flow data found
USGS-04220252 no flow da

KeyError: ignored